tests/rptest/tests/pandaproxy_test.py

```diff
@@ -333,6 +334,127 @@ def _create_named_consumer(self,
+            assert fetch_result_raw.status_code == requests.codes.ok
+            fetch_result_0 = fetch_result_raw.json()
+            assert len(fetch_result_0) == 1
+            assert fetch_result_0[0]["topic"] == topic_name
+            assert fetch_result_0[0]["key"] is None
+            assert fetch_result_0[0]["value"] == expected_data["records"][0][
+                "value"]
+            assert fetch_result_0[0]["partition"] == expected_data["records"][
+                0]["partition"]
+            assert fetch_result_0[0]["offset"] == expected_offset
+
+        def check_offsets(group_id: str,
+                          topic_name: str,
+                          expected_offset: int,
+                          auth_tuple: Optional[tuple[str, str]] = None,
+                          do_set_offsets: bool = True):
+            self.logger.debug(
+                f"Create a consumer and subscribe to topic: {topic_name}")
+            cc_res = self._create_consumer(group_id, auth=auth_tuple)
+            assert cc_res.status_code == requests.codes.ok
+            c0 = Consumer(cc_res.json(), self.logger)
+            sc_res = c0.subscribe([topic_name], auth=auth_tuple)
+            assert sc_res.status_code == requests.codes.no_content
+
+            # Maybe set consumer offsets to 0
+            parts_to_assign = [0, 1, 2]
+            if do_set_offsets:
+                sco_req = dict(partitions=[
+                    dict(topic=topic_name, partition=p, offset=0)
+                    for p in parts_to_assign
+                ])
+                co_res_raw = c0.set_offsets(data=json.dumps(sco_req),
+                                            auth=auth_tuple)
+                assert co_res_raw.status_code == requests.codes.no_content
+
+            self.logger.debug(f"Check consumer offsets")
+            co_req = dict(partitions=[
+                dict(topic=topic_name, partition=p) for p in parts_to_assign
+            ])
+            offset_result_raw = c0.get_offsets(data=json.dumps(co_req),
+                                               auth=auth_tuple)
+            assert offset_result_raw.status_code == requests.codes.ok
+            res = offset_result_raw.json()
+            assert len(res["offsets"]) == len(parts_to_assign)
+            for r in res["offsets"]:
+                assert r["topic"] == topic_name
+                assert r["partition"] in parts_to_assign
+                assert r["offset"] == expected_offset
+                assert r["metadata"] == ""
+
+            # The consumer uses a separate kafka client internal to the http proxy. That client is in a sharded service.
+            # Removing the consumer is necessary to stop the sharded service, or else risk teardown failures.
+            self.logger.info("Remove consumer")
+            rc_res = c0.remove(auth=auth_tuple)
+            assert rc_res.status_code == requests.codes.no_content
```

Comment:
I need to look into this in the C++ code more. I don't think this should be required but, AFAICT, the consumer service is also restarted so this may make sense. Also, I should test if removing the consumer is necessary when we do **not** supply a consumer group. I think the sharded<consumer> is used when a consumer group is passed but that sharded<kafka_client_cache> is used otherwise.

Reply:
> I need to look into this in the C++ code more. I don't think this should be required but, AFAICT, the sharded kafka client is also restarted so this may make sense.

Removing and recreating the consumer is necessary because, when restart is called on the http proxy, the previously create consumer is freed from memory as well. When a consumer is created, we store it in memory in a hash map (see `_consumers` in `kafka/client/client.h`). That in-memory map is auto-destroyed when `stop()` is called on the `sharded<kafka::client>` or `sharded<kafka_client_cache>`. This happens because seastar also frees it's own in-memory vector that contains instances of `kafka::client` or `kafka_client_cache`.